File: Two_pointers/Python/String Compression.py
# Problem link: https://leetcode.com/problems/string-compression/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def compress(self, chars: List[str]) -> int:
        index, count = 0, 1

        for idx in range(1, len(chars) + 1):
            if idx == len(chars) or chars[idx] != chars[idx - 1]:
                chars[index], index = chars[idx - 1], index + 1
                
                if count > 1:
                    for char in str(count):
                        chars[index], index = char, index + 1

                count = 0

            count += 1

        return index

File: Two_pointers/Java/String Compression.java
// Problem link: https://leetcode.com/problems/string-compression/description/
// Time: O(n)
// Space: O(1)
class Solution {
    public int compress(char[] chars) {
        int count = 1, index = 0, size = chars.length;

        for(int idx = 1; idx <= size; idx++) {
            if(idx == size || chars[idx] != chars[idx - 1]) {
                chars[index++] = chars[idx - 1];

                if(count > 1) {
                    for(char chr : String.valueOf(count).toCharArray()) {
                        chars[index++] = chr;
                    }
                }
                count = 0;
            }
            count++;
        }

        return index;
    }
}
